virtualizarr/readers/hdf.py

```diff
@@ -0,0 +1,243 @@
+    dimension scale, it will be considered as the dimension to itself.
+
+    Parameters
+    ----------
+    dataset : h5py.Dataset
+        HDF5 dataset.
+
+    Returns
+    -------
+    list
+        List with HDF5 path names of dimension scales attached to input
+        dataset.
+    """
+    dims = list()
+    rank = len(dataset.shape)
+    if rank:
+        for n in range(rank):
+            num_scales = len(dataset.dims[n])
+            if num_scales == 1:
+                dims.append(dataset.dims[n][0].name[1:])
+            elif h5py.h5ds.is_scale(dataset.id):
+                dims.append(dataset.name[1:])
+            elif num_scales > 1:
+                raise ValueError(
+                    f"{dataset.name}: {len(dataset.dims[n])} "
+                    f"dimension scales attached to dimension #{n}"
+                )
+            elif num_scales == 0:
+                # Some HDF5 files do not have dimension scales.
+                # If this is the case, `num_scales` will be 0.
+                # In this case, we mimic netCDF4 and assign phony dimension names.
+                # See https://github.com/fsspec/kerchunk/issues/41
+                dims.append(f"phony_dim_{n}")
+    return dims
+
+
+def _extract_attrs(h5obj: Union[h5py.Dataset, h5py.Group]):
+    """
+    Extract attributes from an HDF5 group or dataset.
+
+    Parameters
+    ----------
+    h5obj : h5py.Group or h5py.Dataset
+        An HDF5 group or dataset.
+    """
+    _HIDDEN_ATTRS = {
+        "REFERENCE_LIST",
+        "CLASS",
+        "DIMENSION_LIST",
+        "NAME",
+        "_Netcdf4Dimid",
+        "_Netcdf4Coordinates",
+        "_nc3_strict",
+        "_NCProperties",
+    }
+    attrs = {}
+    for n, v in h5obj.attrs.items():
+        if n in _HIDDEN_ATTRS:
+            continue
+        # Fix some attribute values to avoid JSON encoding exceptions...
+        if isinstance(v, bytes):
+            v = v.decode("utf-8") or " "
+        elif isinstance(v, (np.ndarray, np.number, np.bool_)):
+            if v.dtype.kind == "S":
+                v = v.astype(str)
+            if n == "_FillValue":
+                continue
+            elif v.size == 1:
+                v = v.flatten()[0]
+                if isinstance(v, (np.ndarray, np.number, np.bool_)):
+                    v = v.tolist()
+            else:
+                v = v.tolist()
+        elif isinstance(v, h5py._hl.base.Empty):
+            v = ""
+        if v == "DIMENSION_SCALE":
+            continue
+
+        attrs[n] = v
+    return attrs
+
+
+def _dataset_to_variable(path: str, dataset: h5py.Dataset) -> Optional[xr.Variable]:
+    # This chunk determination logic mirrors zarr-python's create
+    # https://github.com/zarr-developers/zarr-python/blob/main/zarr/creation.py#L62-L66
+
+    manifest = _dataset_chunk_manifest(path, dataset)
+    if manifest:
+        chunks = dataset.chunks if dataset.chunks else dataset.shape
+        codecs = codecs_from_dataset(dataset)
```

Comment:
- Given the ZarrV3 spec on codecs being non-empty: https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html#id12
- this comment: https://github.com/zarr-developers/zarr-python/pull/1944#issuecomment-2150056083 mapping `filters` and `compressor` to v3 concepts
- And empirically we observed that on GOES data this builds of a list of [zlib](https://numcodecs.readthedocs.io/en/stable/zlib.html) and [FixedScaleOffset](https://numcodecs.readthedocs.io/en/stable/fixedscaleoffset.html)

Leaving `compressor=None` causes ambiguity for roundtripping v3 metadata (`ZArray -> disk -> ZArray`) because we can't determine if it's a list of 2 filters or a list of one filter and one compressor. `zlib` is a compression codec and `FixedScaleOffset` is not, but should they both be treated as filters?

Reply:
@ghidalgo3 My rationale for describing the full codec chain in the `filters` property was the fact that internally HDF5 does not distinguish compressors and filters, the entire encoding chain is represented as `filters`.  Since we don't need to worry about v2 interoperability, I think we can just focus with aligning with v3's api (which still seem to be in a state of flux).  I think I prefer the approach proposed in https://github.com/zarr-developers/zarr-python/pull/1944#issuecomment-2148231227 but I don't know where that leaves me in the interim until a final decision gets made on the v3 api path 🤔.  For v3 compatibility we'll also need to track https://github.com/zarr-developers/numcodecs/pull/524 so we use `numcodecs` which are compatible with the new v3 codec specification.  TLDR I think we might be in flux for some time while upstream v3 decisions get made.